cpp/include/proxy_client.h

```diff
@@ -98,6 +100,11 @@ class Proxy_client {
   get_object_info(const std::string &namespace_, const std::string &object_name,
                   const consistent_read, const should_cache) = 0;
 
+  virtual void apply_sequence(const std::string &namespace_,
+                              const write_barrier,
+                              const std::vector<std::shared_ptr<sequences::Assert>> &,
+                              const std::vector<std::shared_ptr<sequences::Update>> &) = 0;
+
   /* invalidate_cache influences the result of read requests issued with
```

Comment:
While that interface is definitely sufficient, WDYT about a convenience layer along those lines on top of this (the actual details of the `Assert` and `Update` interfaces can then be hidden):

```
class Assert;
class Update;

BOOLEAN_ENUM(ObjectExists);

class Sequence
{
public:
    using Updates = std::vector<std::unique_ptr<Update>>;
    using Asserts = std::vector<std::unique_ptr<Assert>>;

    Sequence(size_t assert_size_hint = 0, size_t update_size_hint = 0); // hints to reserve vector space
    
    Sequence& add_assert(const std::string&, ObjectExists);
          
    Sequence& add_assert(const std::string&, const std::string& id);

    Sequence& add_assert(const std::string&, const CheckSum&);

    Sequence& add_upload(const std::string&, const std::string& path, const CheckSum*);

    Sequence& add_delete(const std::string&);

private:
    Updates updates_;
    Asserts asserts_;
};
```

and then consequently

```
virtual void apply_sequence(const std::string&, const write_barrier, const Sequence&);
```

Usage example:
```
const auto seq = Sequence().add_assert("foo", ObjectExists::T).add_upload("bar", "/path/to/bar", nullptr);
client->apply_sequence("nspace", write_barrier::T, seq);
```